tests/unit/rpc/WorkQueueTests.cpp

```diff
@@ -99,10 +107,61 @@ TEST_F(WorkQueueTest, NonWhitelistedPreventSchedulingAtQueueLimitExceeded)
         }
     }
 
-    queue.join();
+    queue.stop();
     EXPECT_TRUE(unblocked);
 }
 
+struct WorkQueuePriorityTest : WithPrometheus, RPCWorkQueueTestBase {
+    WorkQueuePriorityTest() : RPCWorkQueueTestBase(1, 100)
+    {
+    }
+};
+
+TEST_F(WorkQueuePriorityTest, HighPriorityTasks)
+{
+    static constexpr auto kTOTAL = 10;
+    std::vector<WorkQueue::Priority> executionOrder;
+    std::mutex mtx;
+
+    for (int i = 0; i < kTOTAL; ++i) {
+        queue.postCoro(
+            [&](auto) {
+                std::lock_guard const lock(mtx);
+                executionOrder.push_back(WorkQueue::Priority::High);
+            },
+            true,
+            WorkQueue::Priority::High
+        );
+        queue.postCoro(
+            [&](auto) {
+                std::lock_guard const lock(mtx);
+                executionOrder.push_back(WorkQueue::Priority::Default);
+            },
+            true,
+            WorkQueue::Priority::Default
+        );
+    }
+
+    queue.stop();
+
+    // with 1 worker, the execution order is deterministic
+    // we should see 4 high prio tasks, then 1 normal prio task, until high prio tasks are depleted
+    std::vector<WorkQueue::Priority> const expectedOrder = {
+        WorkQueue::Priority::High,    WorkQueue::Priority::High,    WorkQueue::Priority::High,
+        WorkQueue::Priority::High,    WorkQueue::Priority::Default, WorkQueue::Priority::High,
+        WorkQueue::Priority::High,    WorkQueue::Priority::High,    WorkQueue::Priority::High,
+        WorkQueue::Priority::Default, WorkQueue::Priority::High,    WorkQueue::Priority::High,
+        WorkQueue::Priority::Default, WorkQueue::Priority::Default, WorkQueue::Priority::Default,
+        WorkQueue::Priority::Default, WorkQueue::Priority::Default, WorkQueue::Priority::Default,
+        WorkQueue::Priority::Default, WorkQueue::Priority::Default,
```

Comment:
Please, write a lambda/func to give this expected order for given amount of high / total number of tasks

Reply:
The whole point is that it's visible in plain text here. writing a lambda means "replicate" the inner algorithm of the work queue - i wanted to avoid that specifically